core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java

```diff
@@ -0,0 +1,247 @@
+ * instance as its ID if the caller so wishes. The consequence of this is that, as a side-effect of
+ * the sorting, the duplicates, as determined by the equality of {@code ID} instances, are removed
+ * from the resulting sorted list.
+ *
+ * @param <N> the type of nodes
+ * @param <ID> the type of a node ID
+ */
+public final class TopologicalSort<N, ID> {
+
+  private final Function<N, ID> identityExtractor;
+  private final Function<N, Set<ID>> directPredecessorsExtractor;
+
+  /**
+   * @param identityExtractor a function to extract some kind of value uniquely identifying a node
+   *     amongst the others.
+   * @param directPredecessorsExtractor a function returning a list of ids of direct predecessors of
+   *     a node
+   */
+  public TopologicalSort(
+      Function<N, ID> identityExtractor, Function<N, Set<ID>> directPredecessorsExtractor) {
+    this.identityExtractor = identityExtractor;
+    this.directPredecessorsExtractor = directPredecessorsExtractor;
+  }
+
+  /**
+   * Given the function for determining the predecessors of the nodes, return the list of the nodes
+   * in topological order. I.e. all predecessors will be placed sooner in the list than their
+   * successors. Note that the input collection is assumed to contain no duplicate entries as
+   * determined by the equality of the {@code ID} type. If such duplicates are present in the input
+   * collection, the output list will only contain the first instance of the duplicates from the
+   * input collection.
+   *
+   * <p>The implemented sort algorithm is stable. If there is no relationship between 2 nodes, they
+   * retain the relative position to each other as they had in the provided collection (e.g. if "a"
+   * preceded "b" in the original collection and there is no relationship between them (as
+   * determined by the predecessor function), the "a" will still precede "b" in the resulting list.
+   * Other nodes may be inserted in between them though in the result).
+   *
+   * <p>The cycles in the graph determined by the predecessor function are ignored and nodes in the
+   * cycle are placed into the output list in the source order.
+   *
+   * @param nodes the collection of nodes
+   * @return the list of nodes sorted in topological order
+   */
+  public List<N> sort(Collection<N> nodes) {
+    // the linked hashmap is important to retain the original order of elements unless required
+    // by the dependencies between nodes
+    LinkedHashMap<ID, NodeInfo<ID, N>> nodeInfos = new LinkedHashMap<>(nodes.size());
+    List<NodeInfo<ID, N>> results = new ArrayList<>(nodes.size());
+
+    int pos = 0;
+    boolean needsSorting = false;
+    for (N node : nodes) {
+      ID nodeID = identityExtractor.apply(node);
+      // we need the set to be modifiable, so let's make our own
+      Set<ID> preds = new HashSet<>(directPredecessorsExtractor.apply(node));
+      needsSorting |= !preds.isEmpty();
```

Comment:
This is not bit manipulation since the variable is a boolean, but I have changed it. I agree your suggestion is easier to read.

Reply:
Oh, I didn't know that this is specified in the language spec https://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.22.2
Thanks for sharing!
But again, this might not be very well known stuff. In a case, when XOR is needed I would go with ^.